sw/host/opentitanlib/src/util/unknown.rs

```diff
@@ -0,0 +1,292 @@
+        #[repr(transparent)]
+        $vis struct $Enum(pub $type);
+
+        #[allow(non_upper_case_globals)]
+        impl $Enum {
+            $(
+                $(#[$inner])*
+                $vis const $enumerator: $Enum = $Enum($value);
+            )*
+        }
+
+        // Implement the various display traits.
+        $crate::__impl_fmt_unknown!(Display, "{}", "{}", $Enum { $($enumerator),* });
+        $crate::__impl_fmt_unknown!(LowerHex, "{:x}", "{:#x}", $Enum { $($enumerator),* });
+        $crate::__impl_fmt_unknown!(UpperHex, "{:X}", "{:#X}", $Enum { $($enumerator),* });
+        $crate::__impl_fmt_unknown!(Octal, "{:o}", "{:#o}", $Enum { $($enumerator),* });
+        $crate::__impl_fmt_unknown!(Binary, "{:b}", "{:#b}", $Enum { $($enumerator),* });
+
+        // Manually implement Serialize and Deserialize to have tight control over how
+        // the struct is serialized.
+        const _: () = {
+            use serde::ser::{Serialize, Serializer};
+            use serde::de::{Deserialize, Deserializer, Error, Visitor};
+            use std::convert::TryFrom;
+
+            impl Serialize for $Enum {
+                /// Serializes the enumerated values.  All named discriminants are
+                /// serialized to strings.  All unknown values are serialized as
+                /// integers.
+                fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+                where
+                    S: Serializer,
+                {
+                    match *self {
+                        $(
+                            $Enum::$enumerator => serializer.serialize_str(stringify!($enumerator)),
+                        )*
+                        $Enum(value) => value.serialize(serializer),
+                    }
+                }
+            }
+
+            // The `EnumVistor` assists in deserializing the value.
+            struct EnumVisitor;
+            impl<'de> Visitor<'de> for EnumVisitor {
+                type Value = $Enum;
+
+                fn expecting(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+                    f.write_str(concat!("A valid enumerator of ", stringify!($Enum)))
+                }
+
+                fn visit_str<E: Error>(self, value: &str) -> Result<Self::Value, E> {
+                    match value {
+                        $(
+                            stringify!($enumerator) => Ok($Enum::$enumerator),
+                        )*
+                        _ => Err(E::custom(format!("unrecognized: {}", value))),
+                    }
+                }
+                fn visit_u64<E: Error>(self, value: u64) -> Result<Self::Value, E> {
+                    match <$type>::try_from(value) {
+                        Ok(v) => Ok($Enum(v)),
+                        Err(_) => Err(E::custom(format!("cannot convert {:?} to {}({})",
+                                      value, stringify!($Enum), stringify!($type)))),
+                    }
+                }
```

Comment:
These visit blocks mostly seem to have the same form, would it make sense to go a level deeper on the macros?
```
macro_rules! expand_visit_func {
    ($visit_func:ident, $type:ty, $Enum:ident) => {
        fn $visit_func<E>(self, value: $type) -> Result<Self::Value, E>
        where
            E: de::Error,
        {
            match <$type>::try_from(value) {
                Ok(v) => Ok($Enum(v)),
                Err(_) => Err(E::custom(format!("cannot convert {:?} to {}({})",
                              value, stringify!($Enum), stringify!($type)))),
            }
        }
    };
}

$crate::expand_visit_func!(visit_u8, u8, $Enum);
$crate::expand_visit_func!(visit_u16, u16, $Enum);
...
```

Reply:
Done